### ISSUE TITLE
Interop needs test with versions before recent refactors

### ISSUE THREAD

**Issue by mlyle** [openLRSng/openLRSng#205]
In this comment, @csurf indicates this tree is intended to keep interop with existing OpenLRSng implementations, while something in the future will be published that isn't compatible:

https://github.com/openLRSng/openLRSng/issues/202#issuecomment-376814758

Interop was recently broken (e.g. #203).  These need to be tested:

* [ ] Old TX, new RX
* [ ] Old RX, new TX.

**Comment by mikeller** [comment]
Happy to do some testing when I find the time.
Also, totally agree that interoperability should be preserved for as much as possible - there will be little interest from existing users to move to new hardware if it means that they have to ditch and replace **all** of their existing hardware at once.

**Comment by Rusty105** [comment]
Can this be bench testing?  Have some spare TX and RX modules to play with and can do some testing.

**Comment by mlyle** [comment]
Yah, bench testing would be great.  I plan to do a test sequence, but I'm having trouble finding my old RXes.

**Comment by mlyle** [comment]
I think 3.8.9 vs. the current code in git would be best.

**Comment by Rusty105** [comment]
OK, I'll see what is loaded on the hardware I have later when I get home. "Current code" being Master or Development?

**Comment by mlyle** [comment]
I think development is abandoned and all the new work has been done in Master.

**Comment by Rusty105** [comment]
ok, I'll check what I already have at home and go from there

**Comment by mikeller** [comment]
@mlyle: Yes, the `development` branch is not used any more, and all development happens on `master`, using the 'cactus model' as postulated in https://barro.github.io/2016/02/a-succesful-git-branching-model-considered-harmful/.

**Comment by mikeller** [comment]
And thanks for testing, @Rusty105.

**Comment by Rusty105** [comment]
@mikeller , no problem, just hope no one is in a hurry....

So the TXs I have
DTFUHF/Hawkeye 1W deluxe on 3.8.6
DTFUHF jr module on 3.8.5

RXs
DTFUHF/Hawkeye 8ch w USB mini  on 3.8.5
(2) DTFUHF/Hawkeye 6ch on ?????  One will connect to both transmitters, one will con connect to either TX. 
I need to get a 3,3 v FDTI adapter to connect to the 2 RXs without the usb port

My Configurator is on 0.57.

Looks like I have some updating to do before i start.

**Comment by mlyle** [comment]
Yes, that's exactly it, whether 3.8.8/3.8.9 can reliably talk to new code, and whether new code can talk to 3.8.8/3.8.9.  THere's already been at least one interop bug introduced (and fixed).

**Comment by Rusty105** [comment]
@mikeller 
AH, got it.....
Yeah I heard about the chrome issue. Good run while it lasted I suppose...... Google getting too big for our own good..  :(

**Comment by csurf** [comment]
Just to chime in and leave my $0.02 (yet at the risk of causing disagreement), I don't see why interop between versions would be an issue. Best practice should be to always run consistent firmware versions across all devices. This avoids introducing any possible issues in general, in addition to possible changes in EEPROM/bind data structures. 

If you don't want to run the most recent firmware version, then simply downgrade both your TX and RX to the desired firmware version. It doesn't seem like a such a big hassle to do so, and guarantees that everything should work consistently.

In general, I think that the versioning really needs to fixed ASAP. There needs to be some kind of urgent / emergency version upgrade to both the firmware and to Configurator. There have been changes to both which can cause havoc if users aren't aware of the changes. I'd vote for pushing master up to 4.0.0 right away. This allows for setting it as a baseline for any major changes (in addition for what's already been changed). 

Configurator really needs to be setup to handle all versions, as well as being able to dynamically pull both older versions and even the most recent 'development' versions built from master, similar to how betaflight does it. However, this is entirely another issue...

**Comment by mlyle** [comment]
It's really unfortunate for people who have lots of receivers, etc, to have to reflash every single thing when they decide to take advantage of a new feature on the tx.

There are version numbers in the binding structure for this reason, and they've avoided being changed, etc.

**Comment by mlyle** [comment]
(It's also unfortunate to have weird cross-project versioning dependencies, as LP, dR, etc, speak the openlrsng protocol).

**Comment by mlyle** [comment]
@csurf ;) I get that maintaining compatibility is hard, even when you were trying to make no protocol changes.

**Comment by csurf** [comment]
@mlyle 
I understand your concerns
keep in mind that this repo/codebase has been basically stagnant for a long time. This is both a good and bad thing; good because it means "things are the same as they've always been" but bad because it hasn't really evolved. I guess it all depends on what the goal is going forward: to grow & change, or to stay the same.

If we were working with better hardware that had tons of resources, then maybe we could make it more robust and have something like a 'legacy mode' that could support older firmware versions, while still having the ability to fast-forward to a new feature set...

All this being said, nothing has really changed in the protocol. The packet structure, timing, bind data, etc. ... they've all basically remained the same. I believe that the problem you identified with the packet header change was the only real major change, but that has been fixed/reverted. Everything else *should be the same (might be wrong about this, but don't think so). There was a change to the TX config data and also some minor changes to the binary protocol, but non of that should affect binding or communication.

**Comment by mikeller** [comment]
@Rusty105: According to Google, Chrome apps never really took off (and looking around it looks to me like they are right with that, apart from 'niche' apps). I can't blame them for not wanting to spend more effort in running a web shop if not enough people are using it.

@csurf: Agreed re the necessity to update the configurator. Unfortunately, because the existing configurator has been specifically designed to take advantage of the Chrome Web store (by using it as the distribution mechanism for the firmware), rewriting it to work as a standalone app requires quite a bit of work, and that comes on top of the work that is required make it build as a standalone app.  Pushing out another version of the Chrome web store configurator that does not contain the hooks for an upgrade to the standalone version will mean we are just throwing the users under the bus, and leaving them with a non-working version of the configurator, and no obvious way to 'fix' it. This isn't a solution.

> If we were working with better hardware that had tons of resources, then maybe we could make it more robust and have something like a 'legacy mode' that could support older firmware versions, while still having the ability to fast-forward to a new feature set...

I think you need to rethink your approach there. You are postulating to abandon backwards compatibility **and** to move to better hardware. This will mean that, in order to be able to use the new hardware (with the non-backwards-compatible protocol), users will have to buy replacements for all of their hardware at the same time. Compared to allowing users to keep using older hardware in combination with the new hardware, your project will have a much smaller chance of success this way.

**Comment by csurf** [comment]
@mikeller 
oops, I wasn't suggesting that.. it was more like wishing / dreaming / thinking out loud. It would be nice to have lots of resources to be able to do more stuff, but unfortunately it's not an option... Obviously, the project is tied to Arduino/AVR hardware, so there's not much that can be done. Moving to new hardware would basically convert it into a completely new project, even if it were compatible with the older AVR hardware/code.

**Comment by mlyle** [comment]
Basically, IMO, it's fine to break backwards compatibility if it's for a good reason, but breaking things for the sake of expedience no es bueno.  (And there's usually no need).

**Comment by mlyle** [comment]
@mikeller -- as I've mentioned before I maintain a STM32 implementation, which now has TX path code and some (backwards compatible) extensions for faster bidir telemetry rates.

Code is here https://github.com/mlyle/dronin/blob/9662937057e959edb42167440c7ecaa504ccc080/flight/PiOS/Common/pios_openlrs.c

It runs on oplink hardware https://www.banggood.com/OpenPilot-OPLINK-MINI-CC3D-REVO-Universal-Transceiver-TX-RX-Module-Integrating-Remote-Controller-p-1000081.html?cur_warehouse=CN or Revolution/Sparky2 FCs, as either an RX or TX.

Next step is adding a RFM22 independence layer so that it can work with LoRa ICs.

**Comment by mikeller** [comment]
@mlyle: Looks like there is some potential for consolidation already. I am still interested in getting an openLRSng SPI RX driver (analogous to https://github.com/betaflight/betaflight/blob/master/src/main/rx/flysky.c) integrated into Betaflight (which could probably be used out of the box on boards like the Sparky2 running Betaflight. It's not a high priority project for me though, and I hope that one day one of the owners of an STM32 openLRSng TX stack will step up and port it to the Betaflight SPI RX API. :wink:

**Comment by Rusty105** [comment]
OK, to add my .02
I think it can be done, move on to more current hardware, and still remain "openLRSng" I'll use Ardupilot as an example, they moved from the APM boards that were using a 8 bit hardware to the Pixhawk, and now several other boards and environments, and kept the same "brand" name. 

No reason it can't be done here. Move to more current hardware in the TX and RX, have a legacy mode in the TX so users don't have to dig older RXs out of their craft, and move forward. OR spent time on the current hardware and squeeze every drop out of it. The guy on RC Groups (Ultimate LRS) is talking a big game using hobbyking 1W modules. Same hardware as openLRSng? Yes? AND full Mavlink telemetry, so it looks like it can be done.
Either way I'll still do some testing and try to help where I can....

**Comment by atomiclama** [comment]
Hi all.
As mentioned above I have rfm22 drivers integrated to the betaflight spi rx stack and a very simple protocol sitting on top of it. It's not openlrs compatible. I've been playing around with reducing the packet structure upping data rates and other things. 

I hook up rfm22 modules to an old CC3D FC and use it as a TX relay. Also hook them up to the FC for the RX. It's working well, I think the last flight was about 1.5K on a 25mW TX setting, no idea if that is good or not but I'm happy with it ;-) 

It supports serial pass through for MSP and the CLI and any of the betaflight telemetry protocols from the FC back to the TX, but as I don't use them I have not made them that robust. 

I'm currently putting together some HW based on the RFM23BP module and probably STM32F3 based chip.
I also have some RFM69 modules for a play with if I get round to it. 

If other devs are interested then it would be good to get together on this, I will be carrying on using betaflight as base and I have no plans for backward compatibility with openlrsng I don't have the HW to test it, but that is just the stack sitting on top of the drivers so would not be so hard if someone wanted to do it.

**Comment by atomiclama** [comment]
Just for giggles here is a couple of pics of my HW
http://www.multi-rotor.co.uk/index.php?action=dlattach;topic=22196.0;attach=24025;image
http://www.multi-rotor.co.uk/index.php?action=dlattach;topic=22196.0;attach=24028;image

**Comment by mikeller** [comment]
@atomiclama: Clearly needs a bigger voltage regulator. :joy:

**Comment by csurf** [comment]
FYI for those interested, found an issue with 3.9.0 repo master TX code not being able to link to previous 3.8.8 RX code. Found this when doing testing for the above serial bridge issue, #206.

3.9.0 RX code is able to link without any issues with 3.8.8 TX code, so the issue appears to be on the TX side. Will have to dig into it further to see what's going on...

FYI, before anyone jumps down my throat about the issue being due to the recent refactors, please keep in mind that the refactoring was aimed primarily at the 'common', RX, and RFM-related code. The TX code basically stayed the same as before except for some minor changes necessary to accommodate the changes to common code. The fact that the latest 3.9.0 RX code still works normally with 3.8.8 should attest to this.

I was working on a big PR to refactor all of the TX code, but decided to pull back on it for personal reasons. So, again, the TX code is pretty much at the 'pre-csurf refactor' state.

The only other major changes that the TX code has undergone since 3.8.8 was the refactoring/restructuring necessary for the MULTI serial protocol support.

For anyone interested in testing this, you'll have to recompile the 3.9.0 code yet setting the version to 3.8.8 (in `version.h`). Otherwise, you won't be able to bind. You should be able to verify that 3.8.8 TX code will bind, link, and function correctly with 3.9.0 RX code (please post back if you find any issues with this).

I have to figure out why the link is dropping between 3.9.0 TX code and 3.8.8 RX code. There might be an issue with the timing, but not sure. Others are welcome & encouraged to test & look into this further. I'd appreciate the help.

**Comment by csurf** [comment]
update on the link/sync issue that I described above:
for some weird reason that I can't quite figure out, there's a timing discrepancy between the latest TX code and the 3.8.8 (and earlier?) RX code. 

In order to make it work, the following change has to be made
https://github.com/openLRSng/openLRSng/blob/master/openLRSng/TX.h#L1029
```
  if ((time - lastSent) >= getInterval(&bind_data) - time_offset) {
```
where `time_offset` is a value somewhere between ~300 and ~1500 microseconds. I'm not quite sure why this is the case. For some strange reason, the packet interval needs to be shortened on the TX side in order for the older firmware to be able to sync up with the TX. 

Unless I'm missing something, it doesn't really make much sense that this should be happening. I don't see anything in the TX code that's changed enough to completely throw off the packet interval timing math. The packet sizes and `getInterval()` function math should be exactly the same as before on both the TX and RX sides.

If anything, I would expect the actual packet interval to be longer on the RX side due to the added latency of processing overhead. However, I don't see why it would be shorter than the interval on TX side for the newer codebase.

In any case, I've stared at this for too long while debugging the serial bridge issue (#206), and would appreciate having a fresh set of eyes take a look at the problem.

**Comment by Rusty105** [comment]
For serial port testing, what is the setup?  Are we just connecting a serial terminal to each end and checking for communications?

**Comment by csurf** [comment]
One thing that I don't believe has been mentioned here (or perhaps I missed it) is that the major and minor version numbers must match in order to bind TX and RX. 

see here:
https://github.com/openLRSng/openLRSng/blob/master/openLRSng/RX.h#L891
https://github.com/openLRSng/openLRSng/blob/master/openLRSng/binding.h#L96


So, backward compatibility can only work across 'patch' versions. We're currently at 3.9.0, so the latest firmware code wouldn't work with anything version prior to 3.9.x anyway... 

Please keep this in mind during any "interop" discussions...

**Comment by mikeller** [comment]
@csurf: Yes, I was indeed not aware of this restriction when I incremented the version number to 3.9.0. Breaking interoperability with this wasn't my intention.

On the other hand, seeing how the version is defined in https://github.com/openLRSng/openLRSng/blob/658e8617331693e0aea297bc990c18c2d9d57cf7/openLRSng/version.h#L5-L6
I do not think we have any other way to increment the version without introducing this break, since only single digit patch version numbers are supported.

As a (decidedly poor) workaround, I can see us create a fork off the 3.8.9 release, increment the version to 3.9.0, and release as a 'interoperability release', with the first 'true' 3.9 release then being 3.9.1. Thoughts @mlyle, @kh4?

**Comment by csurf** [comment]
Well, as the version is in hex/byte values, obviously, there's still the ability to go up to 3.8.f. So, we could probably just downgrade the repo to 3.8.9 or 3.8.a.

However, I still don't think this is such a big deal in general. I don't think it's such ask to require users to occasionally upgrade both their TX and RX gear if they want to run a newer firmware.

**Comment by csurf** [comment]
lol, making the quality of the code worse? As opposed to what? The monolithic rat's nest of spaghetti code that's existed up until now? 

The underlying code structure & algorithms are almost exactly the same as before. The only thing that was done was a *refactoring & clean-up of the existing code in order to break things up into more discrete functions/routines that make more sense logically, which is far better than having one massive function that does everything and that you have to dig into in order to find out how some specific routine works. If you have issues with the 'quality' of the refactored code, then your complaints would also apply to the original code, as they're both basically doing the exact same thing.

Also, you have no proof of anything being "broken" due to your complaints. It's obvious that you have never even tested any of this. This is clearly demonstrated by the fact that you weren't even aware of the 'bind version match' requirement. Unless you manually hacked the firmware version in the code (as I actually did), there's no way you would have even been able to bind the latest repo code to an earlier firmware build in the first place. So, you wouldn't even have been able to test any of your "interop" concerns that you mention here.

The "hop channel in check header" issue was only based on your theoretical assumptions. You didn't even test this with the latest firmware; you just assumed that it wouldn't work. However, I actually sat here with a bound TX-RX setup on my desk, constantly testing the latest firmware, in order to see if there were any link issues. I've literally left the gear running for days on end, frequently power-cycling the TX in order to confirm that the RX could re-bind, which it always did. This was at extremely close range, mind you, where your "hop channel in check header" issue should have surfaced if it was actually a problem (due to the proximity & possibility of "swamping" of the RX).

The fact that you claim that people have 'trouble missing frames' also clearly demonstrates that you don't really understand how this system works (FYI, they're packets, not frames), and I have no idea what you're talking about with the timing being 'tighter'. If anything, the timing was mistakenly "widened" a bit by my offset typo in the getInterval() function (see #211), which would actually make it a bit safer/more conservative in terms of packet timing. 

It looks more likely that the problem is that you simply don't  approve of anyone other than kha making any changes here. Why this is the case is beyond me, but unless you're actually contributing, I don't see how your baseless complaints actually matter.

BTW, what exactly have you contributed to this lately, other than to bitch&complain? I don't see anyone else doing anything here. I've been the only one doing any actual work on the code. 

This is exactly the attitude that has turned me off from contributing to this project; people that don't do anything (and don't really even understand how it works) that come out of the woodwork in order to spout off and whine without actually doing anything productive. 

Prior to @mikeller & myself stepping up to work on the repo, it was basically a dead, stagnant project. No one seemed to care about what was or wasn't done to the code at that point, yet no, all of a sudden, everyone's a concerned "expert"... 🙄

**Comment by mlyle** [comment]
@csurf --- It's nice that you've attempted to factor things more nicely.  But when you reduce the *already very limited headroom* on sync, you make a lost frame much more difficult to recover from.  When you don't ensure channel numbers match before accepting a frame, you will accept frames from adjacent channels and prevent sync.  These are important considerations.   First, do no harm.

**Comment by csurf** [comment]
sigh dude... 
You might want to read it again; the timing was widened, not tightened. It's been put back to the way it was. It was a non-issue in either case (based on real-world testing).

I'm over this. Feel free to keep ranting about "interop" on this issue thread; I'm moving on...

**Comment by mlyle** [comment]
> You might want to read it again; the timing was widened, not tightened.

Yah, increasing the interval and not increasing the corresponding pad value (The other +1000) actually tightens the timing... draw a timing diagram, it's simple.

> It was a non-issue in either case (based on real-world testing).

If you force a lost frame it sure looked like the newer code tends to have to go to slow-hop to resync (and other people have reported this).  Did you inject faults to check the behavior?

**Comment by mikeller** [comment]
@csurf: 

> Well, as the version is in hex/byte values, obviously, there's still the ability to go up to 3.8.f

I am reluctant to use hex version numbers, as this will break places like the configurator that expect version numbers to be SemVer. And breaking SemVer to help us with the current problem, only to then be facing the same problem 6 patch releases later seems to be short sighted.

@mlyle: 

> Asking people to upgrade both occasionally for a good reason is OK.  

Agreed. However, 'we have run out of single digit patch numbers' is **not** a _good_ reason.

Also I don't think that having a fight over the protocol changes that were accidentally introduced makes much sense. They were introduced through no bad intentions (and largely caused by the code currently being hard to understand and barely documented), and have been reverted now. Mistakes happen when things are changed, and I really appreciate @csurf's efforts to improve the code base, and with this make it easier to avoid this kind of mistake in the future.
It is all good again now, and we should move on and focus on things that still need doing.
